### ISSUE TITLE
Rust support

### ISSUE THREAD

**Issue by emoon** [deplinenoise/tundra#261]
Hi,

I would like to add Rust support to tundra but I'm not sure how to go about it. Just adding support for compiling rust files as is very easy as it's just a regular compiler. 

The thing is that Rust has it's own build/package manager called Cargo and some info can be found about it here: http://doc.crates.io/guide.html

The way it works is that you set up a project according to some rules and allows you to depend on packages that can be downloaded when you build and such which is very convenient.

The best way I see it would be if it would be possible to start "cargo build" for a given rust project within tundra and have tundra then link with the output in the cargo output folder (in cargo's case it ends up like target/debug/foo)

Reading the FAQ on cargo it also states this:

#### Can Cargo be used inside of make (or ninja, or ...)
Indeed. While we intend Cargo to be useful as a standalone way to compile Rust projects at the top-level, we know that some people will want to invoke Cargo from other build tools.

We have designed Cargo to work well in those contexts, paying attention to things like error codes and machine-readable output modes. We still have some work to do on those fronts, but using Cargo in the context of conventional scripts is something we designed for from the beginning and will continue to prioritize.

http://doc.crates.io/faq.html

So I feel it should be possible to allow a nice integration into Tundra but I'm unsure of how to do it.

**Comment by emoon** [comment]
Bump

**Comment by deplinenoise** [comment]
Sorry, Rust support seems like very low priority to me so I can't justify spending any time on it at the moment. If you have specific questions on how to make something work (and do most of the research legwork) I can try to help you out.

**Comment by emoon** [comment]
Alright. Thanks!

**Comment by emoon** [comment]
Found another way to handle this so will close this one for now.

**Comment by emoon** [comment]
So, I have decided I need this anyway but I can implement it myself but might a few pointers.

So the way Rust works is that you have a compiler (rustc) which works pretty much like any C/C++ compiler and then you have Cargo which is Rust package manager which download dependencies for your projects, compiles the dependent code and generates the output (which can be a static lib, dynamic lib or program) static libs can be used fine with regular C/C++ code if wanted. My plan is to use Cargo directly as it works nice with the rest of the Rust echo system.

So my idea is to have something like this

```lua
RustStaticLibrary {
  name = "foo",
  Config = "path/to/lib/Cargo.toml", -- Describes depends name, type, etc, of lib
  Sources = { "path/to/lib/src/lib.rs" },
}
```
Example of program

```lua
RustProgram {
  name = "bar",
  Config = "path/to/prog/Cargo.toml", -- Describes depends name, type, etc, of lib
  Sources = { "path/to/prog/src/lib.rs" },
  Depends = { "foo", "some_other_lib" },
}
```
So a bunch of questions:

1. I wonder how tundra should keep track of when a RustProject should be built. Is it possible to have Sources = { ... } as input which will retrigger the building of the project even without actually generating any object file for each file (as cargo is the one that actually calls the rustc compiler)

2. Will it work that other projects depend on the RustProjects (like static libs in the same way?) As a static lib for Rust is just the same thing as a static lib for C/C++ that would be nice to have.

**Comment by deplinenoise** [comment]
I'm not sure how you can teach tundra about the inner workings of some external tool that can download files off the internet? Is maybe this entire thing better expressed inside cargo? Or to rephrase: what problem does Tundra solve in this case, that the rust build system doesn't already solve?

**Comment by emoon** [comment]
Tundra doesn't really need to care about cargo is downloading files from internet. The only thing Tundra needs to know about is when is if user changes tho cargo config or changes some source files it should run the Cargo build step.

Cargo allows you to build C/C++ code but it's fairly limited. Tundra is much better in this regard and I still want to build a bunch of 'external' C/C++ code and link it to some Rust project.

Also Cargo doesn't try to be a generic build system so running steps after everything is built (for example OsxBundle) isn't possible it Cargo. Of course I could do some script that does it every time but that doesn't feel a good long term solution.

Tundras dependency checker is faster that Cargos (which this is being improved it would be nice to use Tundras speed for this)

I hope this makes it a bit more clear :)

**Comment by emoon** [comment]
I have started to work on this now over here https://github.com/emoon/tundra/tree/rust-support

**Comment by emoon** [comment]
I have most of this working now but I have run into one issue. On Mac everything works as expected but on Windows Cargo say 
```
no stdio handle available for this process
```
When looking at the code it seems like GetStdHandle(...) returns NULL.

Do you have any idea why that could be? If I run the exact same commands from the outside (not calling it from Tundra) it seems to work fine.

**Comment by deplinenoise** [comment]
Tundra redirects the file handles so it can capture the console output and present in a nicer way when there are multiple processes all trying to write at the same time.

The relevant code is in `ExecWin32.cpp`:

```c++
  sinfo.hStdInput = NULL;
  sinfo.hStdOutput = sinfo.hStdError = output_handle;
```

Tundra doesn't give the tools a stdin, because it doesn't want them messing with the console. Maybe that's the problem here? Does Cargo want to ask the user questions?? If it does, what happens when you run multiple instances in parallel? You wont know which one you're interacting with :)

**Comment by emoon** [comment]
Yeah that sounds weird. I will take a look at the code :) thanks for the info.

**Comment by emoon** [comment]
So I guess this is the issue (as described in the Cargo source)

```
//!
//! Cargo compile currently does the following steps:
//!
//! All configurations are already injected as environment variables via the
//! main cargo command
//!
//! 1. Read the manifest
//! 2. Shell out to `cargo-resolve` with a list of dependencies and sources as
//!    stdin
//!
//!    a. Shell out to `--do update` and `--do list` for each source
//!    b. Resolve dependencies and return a list of name/version/source
//!
//! 3. Shell out to `--do download` for each source
//! 4. Shell out to `--do get` for each source, and build up the list of paths
//!    to pass to rustc -L
//! 5. Call `cargo-rustc` with the results of the resolver zipped together with
//!    the results of the `get`
//!
//!    a. Topologically sort the dependencies
//!    b. Compile each dependency in order, passing in the -L's pointing at each
//!       previously compiled dependency
//!
```

**Comment by deplinenoise** [comment]
I don't know what they mean by "shelling out", but I assume they're running a subprocess somehow. Maybe they freak out in that code when there's no `stdin` around? But from the description above it sounds like they're trying to set up a new `stdin` for the subprocess, so that seems fine.

If you run cargo with a `< nul` redirection in a regular cmd (essentially closing stdin the same way Tundra does), does that break too?

**Comment by deplinenoise** [comment]
Can you find that exact error message in the cargo code and see why they're freaking out exactly?

**Comment by emoon** [comment]
Building with ```< nul``` works just fine at least.

The error message comes from here: 

https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/stdio.rs#L40

Which is from Rust stdlib when GetStdHandle(...) returns NULL

```Rust
pub fn get(handle: c::DWORD) -> io::Result<Output> {
    let handle = unsafe { c::GetStdHandle(handle) };
    if handle == c::INVALID_HANDLE_VALUE {
        Err(io::Error::last_os_error())
    } else if handle.is_null() {
        Err(io::Error::new(io::ErrorKind::Other,
                           "no stdio handle available for this process"))
    } else {
        let ret = NoClose::new(handle);
        let mut out = 0;
        match unsafe { c::GetConsoleMode(handle, &mut out) } {
            0 => Ok(Output::Pipe(ret)),
            _ => Ok(Output::Console(ret)),
        }
    }
}
```

And I guess Cargo bails if it can't get the handle and and just reports the error it gets from the handler.

**Comment by emoon** [comment]
I will dive more into the Cargo code and try to figure out how it's being used.

**Comment by emoon** [comment]
So I think I found the problem.

https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/process.rs#L147

This code

```Rust
        let stdin = try!(in_handle.to_handle(c::STD_INPUT_HANDLE));
```

Will bail in spawn for a Process and Cargo doesn't set that it doesn't need stdin (from what I understand) so the spawn code defaults to that the user may want to use stdin

**Comment by emoon** [comment]
If I force Stdin to be null in Cargo before spawning a new process it seems to work fine. I will talk with the Cargo devs about it but that seems to be the issue at least.

**Comment by emoon** [comment]
So it seem this caused by a bug in the Rust std lib on Windows (which Cargo uses) https://github.com/rust-lang/rust/issues/31167

I tested this on Mac and Linux and everything works as expected there.

**Comment by emoon** [comment]
Closing this as merged to main repro in https://github.com/deplinenoise/tundra/commit/9f3d161f06ee561ec71122446c512c56ddc31af9
